lib/parser.js

```diff
@@ -0,0 +1,99 @@
+const $RefParser = require('@apidevtools/json-schema-ref-parser');
+const { JSONPath } = require('jsonpath-plus');
+const { takeWhile, merge } = require('lodash');
+
+class ExternalComponents {
+    ref;
+    resolvedJSON;
+    constructor(ref, resolvedJSON) {
+        this.ref = ref;
+        this.resolvedJSON = resolvedJSON;
+    }
+
+    getKey(){
+        const keys = this.ref.split('/');
+        return keys[keys.length -1 ]
+    }
+
+    getValue() {
+        return this.resolvedJSON;
+    }
+}
+
+async function parse(JSONSchema) {
+    const $ref = await $RefParser.resolve(JSONSchema);
+    const refs = crawlChanelPropertiesForRefs(JSONSchema);
+    for (let ref of refs) {
+        if (isExternalReference(ref)) {
+            const componentObject = await resolveExternalRefs(JSONSchema, $ref);
+            merge(JSONSchema.components, componentObject);
+            console.log(JSONSchema.components)
+        }
+    }
+
+
+
+
+
+}
+
+
+function crawlChanelPropertiesForRefs(JSONSchema) {
+    return JSONPath({ json: JSONSchema, path: `$.channels.*.*.message['$ref']` });
+}
+
+
+/**
+ * 
+ * @param {string} ref 
```

Comment:
Would love a comment explaining the function, even though it's pretty self explanatory 🙂